proxy.go

```diff
@@ -413,25 +413,32 @@ func requireBasicAuth(authenticator *auth.BasicAuth, handler func(http.ResponseW
 	})
 }
 
-
 func WithSuitableAuthentication(authenticator *auth.BasicAuth, handler func(http.ResponseWriter, *http.Request)) func(http.ResponseWriter, *http.Request) {
-	if guestAccessAllowed == false {
-		//Return a delegate that always requires basic authentication
-		return requireBasicAuth(authenticator, handler)
-	} else {
-		//Return a delegate that allows request to either be anonymous or correctly authenticated
-		return func(w http.ResponseWriter, r *http.Request) {
-			_,_,basicAuthPresent := r.BasicAuth()
-			if false == basicAuthPresent {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if !guestAccessAllowed {
+			//All requests require authentication
+			delegate := requireBasicAuth(authenticator, handler)
+			delegate(w, r)
+		} else {
+			_, _, basicAuthPresent := r.BasicAuth()
+			if !basicAuthPresent {
 				//Run the handler as unauthenticated user
-				handler(w,r)
+				confLock.RLock()
+				if _, guestQuotaExists := quota[guestUserName]; !guestQuotaExists {
+					//Return 401
+					reply(w, errMsg("Guest access is unavailable."), http.StatusUnauthorized)
+
+				} else {
+					handler(w, r)
+				}
+				confLock.RUnlock()
```

Comment:
@adrichem after some discussion, please leave locking as is. Only change `isGuestUserPresent` -> `ok`.